clients/drcachesim/tools/schedule_stats.cpp

```diff
@@ -0,0 +1,295 @@
+{
+    auto per_shard = new per_shard_t;
+    std::lock_guard<std::mutex> guard(shard_map_mutex_);
+    per_shard->stream = stream;
+    per_shard->core = stream->get_output_cpuid();
+    shard_map_[shard_index] = per_shard;
+    return reinterpret_cast<void *>(per_shard);
+}
+
+bool
+schedule_stats_t::parallel_shard_exit(void *shard_data)
+{
+    // Nothing (we read the shard data in print_results).
+    return true;
+}
+
+std::string
+schedule_stats_t::parallel_shard_error(void *shard_data)
+{
+    per_shard_t *per_shard = reinterpret_cast<per_shard_t *>(shard_data);
+    return per_shard->error;
+}
+
+bool
+schedule_stats_t::parallel_shard_memref(void *shard_data, const memref_t &memref)
+{
+    per_shard_t *shard = reinterpret_cast<per_shard_t *>(shard_data);
+    if (knob_verbose_ >= 4) {
+        std::ostringstream line;
+        line << "Core #" << std::setw(2) << shard->core << " @" << std::setw(9)
+             << shard->stream->get_record_ordinal() << " refs, " << std::setw(9)
+             << shard->stream->get_instruction_ordinal() << " instrs: input "
+             << std::setw(4) << shard->stream->get_input_id() << " @" << std::setw(9)
+             << shard->stream->get_input_interface()->get_record_ordinal() << " refs, "
+             << std::setw(9)
+             << shard->stream->get_input_interface()->get_instruction_ordinal()
+             << " instrs: " << std::setw(16) << trace_type_names[memref.marker.type];
+        if (type_is_instr(memref.instr.type))
+            line << " pc=" << std::hex << memref.instr.addr << std::dec;
+        else if (memref.marker.type == TRACE_TYPE_MARKER) {
+            line << " " << memref.marker.marker_type
+                 << " val=" << memref.marker.marker_value;
+        }
+        line << "\n";
+        std::cerr << line.str();
+    }
+    if (memref.marker.type == TRACE_TYPE_MARKER &&
+        memref.marker.marker_type == TRACE_MARKER_TYPE_CORE_WAIT) {
+        ++shard->counters.waits;
+        if (!shard->prev_was_wait) {
+            shard->thread_sequence += '-';
+            shard->cur_segment_instrs = 0;
+            shard->prev_was_wait = true;
+        } else {
+            ++shard->cur_segment_instrs;
+            if (shard->cur_segment_instrs == knob_print_every_) {
+                shard->thread_sequence += '-';
+            }
+        }
+        return true;
+    } else
+        shard->prev_was_wait = false;
+    int64_t input = shard->stream->get_input_id();
+    if (input != shard->prev_input) {
+        // We convert to letters which only works well for <=26 inputs.
+        if (!shard->thread_sequence.empty()) {
+            ++shard->counters.total_switches;
+            if (shard->saw_maybe_blocking || shard->saw_exit)
+                ++shard->counters.voluntary_switches;
+            if (shard->direct_switch_target == memref.marker.tid)
+                ++shard->counters.direct_switches;
+            shard->thread_sequence += ',';
```

Comment:
We add a comma at each switch? Why? isn't the character changing enough? Or do you want to differentiate between thread 0 and thread 26.

Reply:
It seemed easier to read at a glance how long the sequences were.  Compare:
```
Core #0 schedule: AA,CC,HH,LL,UU,WW,YY,AA,CC,EE,FF,GG,HH,II,LL,MMM,KK,EE,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC,KK,CC
Core #0 schedule: AACCHHLLUUWWYYAACCEEFFGGHHIILLMMMKKEEKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCCKKCC
```
Or:
```
Core #0 schedule: FFFFFF,HHHHHHHHHHHHHHHHHHH,FFF
Core #1 schedule: DDDDDDDDDDDDDDDDDD,CCCCCCCCCCCCCCCCCCC
Core #0 schedule: FFFFFFHHHHHHHHHHHHHHHHHHHFFF
Core #1 schedule: DDDDDDDDDDDDDDDDDDCCCCCCCCCCCCCCCCCCC
```

I imagine how we visualize longer sequences will evolve over time though.  I think starting with the , is nice.  I'll add a comment about maybe removing to make it more compact, but that making it harder to distinguish a 26-rollover switch like you said.